src/include/OSL/device_string.h

```diff
@@ -28,21 +28,42 @@ OSL_NAMESPACE_ENTER
 // stored in the table, in the 16 bytes preceding the characters.
 
 struct DeviceString {
+#if defined(__CUDA_ARCH__) && OPTIX_VERSION >= 70000
+    size_t      m_chars;
+#else
     const char* m_chars;
+#endif
+
+#if defined(__CUDA_ARCH__) && OPTIX_VERSION >= 70000
+    OSL_HOSTDEVICE DeviceString() {}
+    OSL_HOSTDEVICE DeviceString(uint64_t i) : m_chars(i) {}
+#endif
 
     OSL_HOSTDEVICE uint64_t hash() const
     {
+#if defined(__CUDA_ARCH__) && OPTIX_VERSION >= 70000
+        return m_chars;
+#else
         return *(uint64_t*)(m_chars - sizeof(uint64_t) - sizeof(uint64_t));
+#endif
     }
 
     OSL_HOSTDEVICE uint64_t length() const
     {
+#if defined(__CUDA_ARCH__) && OPTIX_VERSION >= 70000
+        return 0;
+#else
         return *(uint64_t*)(m_chars - sizeof(uint64_t));
+#endif
     }
 
     OSL_HOSTDEVICE const char* c_str() const
     {
+#if defined(__CUDA_ARCH__) && OPTIX_VERSION >= 70000
+        return nullptr;
+#else
         return m_chars;
+#endif
```

Comment:
I wonder if the better way to handle this (and length()) might be as:

    #if ! (defined(__CUDA_ARCH__) && OPTIX_VERSION >= 70000)
        OSL_HOSTDEVICE const char* c_str() const
        {
            return m_chars;
        }
    #endif 

... and then it's a compile-time error if you try to use c_str() or length() from Cuda+Optix7 code at all? Rather than returning 0/null and possibly having downstream logic errors?

Reply:
I can look into that.